filebeat/input/file/identifier.go

```diff
@@ -0,0 +1,127 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package file
+
+import (
+	"fmt"
+	"io/ioutil"
+	"strconv"
+	"strings"
+
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type StateIdentifier interface {
+	// ID generates and returns the ID of the state
+	ID(*State) string
+	// Identifier returns data to be persisted in the registry.
+	DataToSerialize() interface{}
+}
+
+type inodeDeviceIdentifier struct {
+	name string
+}
+
+func (i *inodeDeviceIdentifier) ID(s *State) string {
+	if s.Id == "" {
+		stateID := i.name + ":" + s.FileStateOS.String()
+		if len(s.Meta) == 0 {
+			s.Id = stateID
+		} else {
+			s.Id = genIDWithHash(s.Meta, stateID)
+		}
+	}
+
+	return s.Id
+}
+
+func (i *inodeDeviceIdentifier) DataToSerialize() interface{} { return nil }
+
+type pathIdentifier struct {
+	name string
+}
+
+func (p *pathIdentifier) ID(s *State) string {
+	if s.Id == "" {
+		stateID := p.name + ":" + s.Source
+		if len(s.Meta) == 0 {
+			s.Id = stateID
+		} else {
+			s.Id = genIDWithHash(s.Meta, stateID)
+		}
+	}
+
+	return s.Id
+}
+
+func (p *pathIdentifier) DataToSerialize() interface{} { return nil }
+
+type inodeMarkerIdentifier struct {
+	log        *logp.Logger
+	name       string
+	markerPath string
+}
+
+func (i *inodeMarkerIdentifier) markerContents() string {
+	if i.markerPath == "" {
+		return ""
+	}
+
+	deviceIDBytes, err := ioutil.ReadFile(i.markerPath)
```

Comment:
this information is not cached and re-read on every single update. When collecting many files this might become quite expensive. Is it possible to read/cache the content of the marker path early. This would also giving us some guarantees that we won't run into IO issue when we try serialize the state.